instance/models/appserver.py

```diff
@@ -0,0 +1,254 @@
+from .utils import ModelResourceStateDescriptor, ResourceState, ValidateModelMixin
+
+
+# Logging #####################################################################
+
+logger = logging.getLogger(__name__)
+
+
+# States ######################################################################
+
+class AppServerState(ResourceState):
+    """
+    A [finite state machine] state describing an instance.
+    """
+    # A state is "steady" if we don't expect it to change.
+    # This information can be used to:
+    # - delay execution of an operation until the target server reaches a steady state
+    # - raise an exception when trying to schedule an operation that depends on a state change
+    #   while the target server is in a steady state
+    # Steady states include:
+    # - Status.New
+    # - Status.Running
+    # - Status.ConfigurationFailed
+    # - Status.Error
+    # - Status.Terminated
+    is_steady_state = True
+
+    # An instance is healthy if it is part of a normal (expected) workflow.
+    # This information can be used to detect problems and highlight them in the UI or notify users.
+    # Healthy states include:
+    # - Status.New
+    # - Status.WaitingForServer
+    # - Status.ConfiguringServer
+    # - Status.Running
+    # - Status.Terminated
+    is_healthy_state = True
+
+
+class Status(ResourceState.Enum):
+    """
+    The states that an instance can be in.
+    """
+
+    class New(AppServerState):
+        """ Newly created """
+        state_id = 'new'
+
+    class WaitingForServer(AppServerState):
+        """ Server not yet accessible """
+        state_id = 'waiting'
+        name = 'Waiting for server'
+        is_steady_state = False
+
+    class ConfiguringServer(AppServerState):
+        """ Running Ansible playbooks on server """
+        state_id = 'configuring'
+        name = 'Configuring server'
+        is_steady_state = False
+
+    class Running(AppServerState):
+        """ Instance is up and running """
+        state_id = 'running'
+
+    class ConfigurationFailed(AppServerState):
+        """ Instance was not configured successfully (but may be partially online) """
+        state_id = 'failed'
+        name = 'Configuration failed'
+        is_healthy_state = False
+
+    class Error(AppServerState):
+        """ Instance never got up and running (something went wrong when trying to build new VM) """
+        state_id = 'error'
+        is_healthy_state = False
+
+    class Terminated(AppServerState):
+        """ Instance was running successfully and has been shut down """
+        state_id = 'terminated'
+
+
+# Models ######################################################################
+
+
+class AppServer(ValidateModelMixin, TimeStampedModel):
+    """
+    AppServer - One or more distinct web applications running on a single VM.
+
+    Owned by an Instance.
+
+    Characteristics of an AppServer:
+    * An AppServer instance's configuration fields are *immutable*. If you want to change
```

Comment:
Sure. That's another reason I didn't like the name "Instance" btw - because it is a useful term when describing an object in an object oriented programming language like Python. e.g. we have to use built-ins `isinstance()` etc.